### PR TITLE
Fixed and improved ./configure on Windows

### PR DESCRIPTION
**Pull request by meteorcloudy**
1. Make the script work with path containing spaces (If users install python by Anaconda, the python path will probably contain space)
2. On Windows, disable bazel clean --expunge(it's broken)
3. Create junctions instead of copying python lib, include, numpy folders
4. fixed symlink creation in gen_git_source.py with python 3.5

### CODE REVIEW

File: tensorflow/tools/git/gen_git_source.py [hunk 1]
```diff
@@ -112,9 +112,14 @@ def configure(src_base_path, debug=False):
     if src is None:
       open(os.path.join(gen_path, target), "w").write("")
     else:
-      if hasattr(os, 'symlink'):
-        os.symlink(src, os.path.join(gen_path, target))
-      else:
+      try:
+        # In python 3.5, symlink function exists even on Windows
+        # But requires previliaged right, otherwise a OSError will be thrown
```
Comment (rohan100jain) [inline tensorflow/tools/git/gen_git_source.py]: s/priviliaged/privileged 

Also can you expand on what privileges are needed?

**Review by rohan100jain — COMMENTED** [review]
You seem to have added a lot of quotes around variables etc. Can you summarize why that is necessary for windows?

File: configure [hunk 2]
```diff
@@ -8,8 +8,22 @@ pushd `dirname $0` #> /dev/null
 SOURCE_BASE_DIR=`pwd -P`
```
Comment (mrry) [inline configure]: Could this include spaces? Do we need quotes around each use of `$SOURCE_BASE_DIR` as well?

File: tensorflow/tools/git/gen_git_source.py (hunk 1 above)
Comment (mrry) [inline tensorflow/tools/git/gen_git_source.py]: IIRC, Windows requires a special user right for creating symbolic links. It's available if you have Administrator privileges, and although it can be extended to regular users, it isn't by default....

File: util/python/python_config.sh [hunk 3]
```diff
@@ -108,26 +108,27 @@ function setup_python {
...
-  local python_include=$("${PYTHON_BIN_PATH}" -c 'from __future__ import print_function; from distutils import sysconfig; print(sysconfig.get_python_inc());')
+  local python_include="$("${PYTHON_BIN_PATH}" -c 'from __future__ import print_function; from distutils import sysconfig; print(sysconfig.get_python_inc());')"
   if [ "$python_include" == "" ]; then
     echo -e "\n\nERROR: Problem getting python include path.  Is distutils installed?"
     exit 1
   fi
-
-  local python_lib_path=$(python_path)
+  local python_lib_path
+  # Split python_path into an array of paths, this allows path cotaining spaces
```
Comment (mrry) [inline util/python/python_config.sh]: Nit: s/cotaining/containing/

File: util/python/python_config.sh [hunk 4]
```diff
@@ -142,24 +143,29 @@ function setup_python {
...
       rm -rf "$x"
     fi
   done
 
-# ln -sf is acutally implemented as copying in msys since creating symbolic links is privileged on Windows
-# So we need -rf to remove them above.
-  ln -sf "${python_include}" util/python/python_include
-  ln -sf "${python_lib}" util/python/python_lib
-  ln -sf "${numpy_include}" third_party/py/numpy/numpy_include
+# ln -sf is acutally implemented as copying in msys since creating symbolic links is privileged on Windows.
```
Comment (mrry) [inline util/python/python_config.sh]: Nit: 80 characters, s/acutally/actually/.

File: util/python/python_config.sh [hunk 5]
```diff
@@ -142,24 +143,29 @@ function setup_python {
...
+# But copying is too slow, so invoke mklink to create junctions on Windows.
+  if is_windows; then
+    cmd  /c "mklink /J util\\python\\python_include \"${python_include}\""
```
Comment (mrry) [inline util/python/python_config.sh]: Nit: one space between `cmd` and `/c` (same on two following lines)

File: util/python/python_config.sh [hunk 6]
```diff
@@ -183,7 +189,12 @@ function check_python {
       echo -e "\n\nERROR: '${x}' is not a symbolic link.  Internal error.\n\n" 1>&2
       exit 1
     fi
-    true_path=$(readlink "${x}")
+    if is_windows; then
+      # In msys, readlink path just doesn't work, because no symbolic link on Windows
```
Comment (mrry) [inline util/python/python_config.sh]: Nit: 80 characters per line.

**Review by mrry — CHANGES_REQUESTED** [review]
With the `is_windows()` function, would it be worth guarding the other configuration options that aren't currently supported on Windows (e.g. maybe GCP, HDFS, CUDA)?

File: configure (hunk 2 above)
Comment (meteorcloudy) [inline configure]: Yes, I believe so.

**Comment by meteorcloudy** [comment]
@mrry Yes, I think it's a good idea to use `is_windows` to opt out GCP, HDFS, CUDA support on Windows for now.

**Comment by meteorcloudy** [comment]
@rohan100jain For example, if users install python using Anaconda(which is recommended in CMake build), the python path will probably contain spaces like this `C:\Program Files\Anaconda3\lib\site-packages`, so we have to add quotes to the variable.

**Comment by meteorcloudy** [comment]
@mrry Thanks for catching so many typos.. Please review again. :)

**Review by mrry — APPROVED** [review]
Looks good to me!

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by rohan100jain** [comment]
@tensorflow-jenkins test this please

**Comment by meteorcloudy** [comment]
The configure script does cause a bug on Mac, I'll look into it.

**Comment by meteorcloudy** [comment]
`read -r -a python_lib_path <<< $(python_path)` doesn't really split the string into a bash array on OSX, so switch to another way.
Please test again! :)

**Comment by DevinCheung** [comment]
I install the tensorflow on windows via the cmake/README. It seems everything goes well. And after I "activate tensorflow" and "import tensorflow as tf". It says "No module named 'tensorflow'" . I do not know how to fix it.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by meteorcloudy** [comment]
Great, looks like I fixed it!
